src/acvp.c

```diff
@@ -1196,6 +1195,402 @@ static ACVP_RESULT acvp_add_drbg_prereq_val(ACVP_DRBG_CAP_MODE *drbg_cap_mode,
 }
 
 /*
+ * Add top level RSA keygen parameters
+ */
+static ACVP_RESULT acvp_add_rsa_keygen_parm (
+                             ACVP_RSA_CAP_MODE_LIST  *rsa_cap_mode_list,
+                             ACVP_RSA_PARM       param,
+                             int                  value
+                             )
+{
+    switch (param) {
+    case ACVP_PUB_EXP:
+        rsa_cap_mode_list->cap_mode_attrs.keygen->pub_exp = value;
+        break;
+    case ACVP_RAND_PQ:
+        rsa_cap_mode_list->cap_mode_attrs.keygen->rand_pq = value;
+        break;
+    case ACVP_RSA_INFO_GEN_BY_SERVER:
+        rsa_cap_mode_list->cap_mode_attrs.keygen->info_gen_by_server = value;
+        break;
+    default:
+        return ACVP_INVALID_ARG;
+        break;
+    }
+
+    return ACVP_SUCCESS;
+}
+
+/*
+ * Append a RSA pre req val to the list of prereqs
+ */
+static ACVP_RESULT acvp_add_rsa_prereq_val(ACVP_RSA_CAP *rsa_cap, ACVP_RSA_PRE_REQ pre_req, char *value)
+{
+    ACVP_RSA_PREREQ_VALS *prereq_entry, *prereq_entry_2;
+
+    prereq_entry = calloc(1, sizeof(ACVP_RSA_PREREQ_VALS));
+    if (!prereq_entry) {
+        return ACVP_MALLOC_FAIL;
+    }
+    prereq_entry->prereq_alg_val.alg = pre_req;
+    prereq_entry->prereq_alg_val.val = value;
+
+    /*
+     * 1st entry
+     */
+    if (!rsa_cap->prereq_vals) {
+        rsa_cap->prereq_vals= prereq_entry;
+    } else {
+        /*
+         * append to the last in the list
+         */
+        prereq_entry_2 = rsa_cap->prereq_vals;
+        while (prereq_entry_2->next) {
+            prereq_entry_2 = prereq_entry_2->next;
+        }
+        prereq_entry_2->next = prereq_entry;
+    }
+    return (ACVP_SUCCESS);
+}
+
+ACVP_RESULT acvp_rsa_prepare_to_add_param(ACVP_CTX *ctx, ACVP_CIPHER cipher,
+                                          ACVP_RSA_MODE mode,
+                                          ACVP_CAPS_LIST **cap_list,
+                                          ACVP_RSA_CAP_MODE_LIST      **rsa_cap_mode_list) {
+
+    ACVP_RSA_CAP_MODE_LIST *current_rsa_cap_list;
+
+    /*
+     * Validate input
+     */
+    if (!ctx) {
+        return ACVP_INVALID_ARG;
+    }
+
+    switch (cipher) {
```

Comment:
Are there more plans for this check or would a simple if (cipher != ACVP_RSA) suffice ?